### PR TITLE
Add kafka tls authentication to settings

### PR DESCRIPTION
**Pull request by alexef**
Fixes: #974. Related to: https://github.com/getsentry/sentry/pull/21275

Add support for additional configuration (like sasl mechanisms and more) to the Snuba settings.

```python
BBROKER_CONFIG: Mapping[str, Any] = {
    "bootstrap.servers": os.environ.get("DEFAULT_BROKERS", "localhost:9092"),
    "sasl.mechanism": os.environ.get("SASL_MECHANISM"),
    "sasl.username": os.environ.get("SASL_PLAIN_USERNAME"),
    "sasl.password": os.environ.get("SASL_PLAIN_PASSWORD"),
    "security.protocol": os.environ.get("SECURITY_PROTOCOL"),
}
STORAGE_BROKER_CONFIG: Mapping[str, Mapping[str, Any]] = {}
```

Note that even there is a single mapping of configuration, one can still use multiple brokers for bootstrapping communication with kafka.

This is in line with how [Sentry itself](https://github.com/getsentry/sentry/blob/4389bf5a4a7421d5ac3322c7f27c695d8fcef1cb/src/sentry/conf/server.py#L1853-L1859) configures the connections to kafka.

### CODE REVIEW

File: snuba/utils/streams/backends/kafka.py [hunk 1]
```diff
@@ -637,6 +637,31 @@ def build_kafka_consumer_configuration(
     }
 
 
+def build_kafka_configuration(bootstrap_servers: Sequence[str],) -> Mapping[str, Any]:
+    return {
+        "bootstrap.servers": ",".join(bootstrap_servers),
```
Comment (olliezhu) [inline snuba/utils/streams/backends/kafka.py:642]: according to [this](https://docs.confluent.io/current/clients/confluent-kafka-python/#pythonclient-configuration), 
> The supported configuration values are dictated by the underlying librdkafka C library
which [supports](https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md)

```python
        ...
        "security.protocl": "sasl_ssl",
    }
```

**Comment by olliezhu** [comment]
Do there need to be similar changes to support SAML and SSL on sentry itself?
- [Related issue?](https://github.com/getsentry/sentry/issues/14123)
- [No `saml_ssl` configured on the `confluent_kafka` `Consumer`](https://github.com/getsentry/sentry/blob/master/src/sentry/eventstream/kafka/consumer.py#L65-L74)

File: snuba/utils/streams/backends/kafka.py (hunk 1 above)
Comment (alexef) [reply snuba/utils/streams/backends/kafka.py]: Hi @olliezhu, thank you for your review.

Please note this commit was an intermediary step in the direction of refactoring the code to support more meaningful configuration. 

The end goal is allowing the librdkafka supported configuration values to be set in the `settings.py`. 

I hope this answers your comment.

**Comment by alexef** [comment]
@olliezhu I'm not yet familiar with how Sentry interacts with kafka. But from what I can tell from here: https://github.com/getsentry/sentry/blob/4389bf5a4a7421d5ac3322c7f27c695d8fcef1cb/src/sentry/conf/server.py#L1853 configuring additional librdkafka settings is already possible in sentry.

**Comment by alexef** [comment]
@olliezhu I see what you mean now. Sentry has _incomplete_ support for additional configuration. On your example, the root cause is here: https://github.com/getsentry/sentry/blob/d46213a3cb4fea1653724ac58bcd78e32a3dda05/src/sentry/eventstream/kafka/backend.py#L87 where instead of passing the entire cluster config, only the `bootstrap.servers` value is passed to the SynchronizedConsumer.

Assuming the approach in this PR is okay, I can give it a try and fix it in sentry as well.

File: snuba/cli/subscriptions.py [hunk 2]
```diff
@@ -138,14 +142,12 @@ def subscriptions(
         SynchronizedConsumer(
             KafkaConsumer(
                 build_kafka_consumer_configuration(
-                    bootstrap_servers,
-                    consumer_group,
-                    auto_offset_reset=auto_offset_reset,
+                    broker_config, consumer_group, auto_offset_reset=auto_offset_reset,
```
Comment (fpacifici) [inline snuba/cli/subscriptions.py:145]: Did you skip the commit hooks? there should be a commit hook that ensure black style is applied. If not could you please ensure you ran `make setup-git` as explained here: https://github.com/getsentry/snuba#install--run-only-required-if-you-are-developing-against-snuba

File: snuba/utils/streams/backends/kafka.py [hunk 3]
```diff
@@ -637,6 +643,25 @@ def build_kafka_consumer_configuration(
     }
 
 
+def build_kafka_producer_configuration(
+    broker_config: Mapping[str, Any],
+) -> Mapping[str, Any]:
```
Comment (fpacifici) [inline snuba/utils/streams/backends/kafka.py:648]: We are trying to fully type hint snuba. Which means we would prefer dataclasses with a well defined list of attributes and their types to move around the code base. This would make the code that depends on the configuration harder to break. 
So could you please declare a dataclass (like `KafkaBrokerConfiguration`) and return an instance here that is built from the settings dictionary ? (better to keep settings as simple dictionaries for now).

Basically like we did for clusters (https://github.com/getsentry/snuba/blob/master/snuba/clusters/cluster.py#L273). Those were not datatclasses but it does not apply here.

**Review by fpacifici — CHANGES_REQUESTED** [review]
Thanks a lot for the contribution. This is appreciated.
Some comments inline that we need to address in order to move on.

File: snuba/cli/subscriptions.py (hunk 2 above)
Comment (alexef) [reply snuba/cli/subscriptions.py:145]: I have the commit hooks installed, did not skip them. Let me call the hook manually to refactor it.

Comment (alexef) [reply snuba/cli/subscriptions.py:145]: Called ```black snuba/cli/subscriptions.py``` and it only removed the trailing comma.

File: snuba/settings.py [hunk 4]
```diff
@@ -69,8 +69,17 @@
...
+    "bootstrap.servers": os.environ.get("DEFAULT_BROKERS", "localhost:9092"),
+    "sasl.mechanism": os.environ.get("SASL_MECHANISM"),
+    "sasl.username": os.environ.get("SASL_PLAIN_USERNAME"),
+    "sasl.password": os.environ.get("SASL_PLAIN_PASSWORD"),
+    "security.protocol": os.environ.get("SECURITY_PROTOCOL"),
+}
+DEFAULT_BROKERS: Mapping[str, Any] = {
+    k: v for k, v in _DEFAULT_BROKERS.items() if v is not None
+}
+DEFAULT_STORAGE_BROKERS: Mapping[str, Mapping[str, Any]] = {}
```
[the remark this answers is not in the document]
Comment (alexef) [reply snuba/settings.py]: Thanks @fpacifici. I've pushed now a change that preserves the original config file and existing DEFAULT_STORAGE_BROKERS settings should continue to work.

In places where this is used, I'm checking if a DEFAULT_STORAGE_BROKER exists for a given storage_key. If not, I'm using the newly introduced STORAGE_BROKER_CONFIG mapping.

This means an user with a custom settings_prod and both DEFAULT_STORAGE_BROKER (old way) and STORAGE_BROKER_CONFIG (new way) will get the old way config - just bootstrap.servers. Should I prefer the new way instead? Or print a warning that the old way is being used, and they should just kill their custom DEFAULT_STORAGE_BROKER configuration in favor of the new STORAGE_BROKER_CONFIG?

Happy to re-do it.

File: snuba/utils/streams/backends/kafka.py (hunk 3 above)
Comment (alexef) [reply snuba/utils/streams/backends/kafka.py:648]: I see your point, but the return value of `build_kafka_consumer_configuration`, `build_kafka_producer_configuration` and `build_admin_client_configuration` is directly passed to the confluent-kafka classes: Consumer, Producer and AdminClient.

I can make these function return a data class, only to immediately convert the data class back to a dictionary, when instantiating the library classes.

Would this be better than the current approach?

Comment (alexef) [reply snuba/utils/streams/backends/kafka.py]: Updated the code to reflect the above comment in https://github.com/getsentry/snuba/pull/1406/commits/778458fe183829d2dc80ef661c9c1a5eb0ba528e. I'm not really happy with how it looks. wdyt?

File: snuba/cli/subscriptions.py (hunk 2 above)
Comment (fpacifici) [reply snuba/cli/subscriptions.py]: You are right, you made the statement just short enough that it fits on a line. Please ignore the comment.

File: snuba/utils/streams/backends/kafka.py [hunk 5]
```diff
@@ -617,24 +619,55 @@ def closed(self) -> bool:
 DEFAULT_QUEUED_MIN_MESSAGES = 10000
 
 
+@dataclass
+class KafkaBrokerConfig:
+    config: Mapping[str, Any]
```
Comment (fpacifici) [inline snuba/utils/streams/backends/kafka.py:624]: I originally meant a dataclass with each field we allow to be explicitly declared (so we would have a list of allowed config parameters), but I did not realize that we do absolutely nothing with that except passing it to the kafka api, so you are right, that was a bad idea. Sorry about that.

So I would go for the original idea of using a dictionary and we can turn into a typed dict later. Also we do not need to have a list of allowed fields.

File: snuba/utils/streams/backends/kafka.py [hunk 6]
```diff
@@ -617,24 +619,55 @@ def closed(self) -> bool:
...
+
+
+def get_broker_config(bootstrap_servers: Sequence[str]) -> Mapping[str, Any]:
+    return {
+        "bootstrap.servers": ",".join(bootstrap_servers),
+    }
+
+
 def build_kafka_consumer_configuration(
```
Comment (fpacifici) [inline snuba/utils/streams/backends/kafka.py:633]: Is there any reason not to encapsulate the logic that loads the settings here? 
It seems that logic is spread over multiple scripts. What about having these methods here:
```
# This looks for the storage annd then falls back to the default
def build_kafka_consumer_config(storage_name: str, group_id: str ... , override_params: Mapping[str, Any]) -> Mapping [str: Any]

# This does not take in the storage
def build_default_kafka_consumer_config(group_id: str ... , override_params: Mapping[str, Any]) -> Mapping [str: Any]

def build_producer_config(storage: str, override_params: Mapping[str, Any]) -> Mapping[str: Any]
```

These methods would load the config, apply the overrides (ensuring we do not try to override the server name for integrity of our config), remove the None values, and return the packed dictionary.

Also I would add a type alias for the returned field:
`KafkaBrokerConfig = Mapping[str, Any]`

It is mostly cosmetics in this file but it makes every file depending on this referencing a specific type that conveys something about its semantics instead of being a bare mapping.

File: snuba/utils/streams/backends/kafka.py [hunk 7]
```diff
@@ -617,24 +619,55 @@ def closed(self) -> bool:
...
+
+
+def build_kafka_producer_configuration(
+    broker_config: Mapping[str, Any],
+) -> KafkaBrokerConfig:
+    return KafkaBrokerConfig(
+        {
+            **broker_config,
+            "partitioner": "consistent",
+            "message.max.bytes": 50000000,  # 50MB, default is 1MB
```
Comment (fpacifici) [inline snuba/utils/streams/backends/kafka.py:662]: See my comment above, I would not hardcode parameters in this code and allow the methods to receive an override mapping.

**Review by fpacifici — CHANGES_REQUESTED** [review]
This is looking good, thanks. 
Some additional requests.

File: snuba/settings.py [hunk 8]
```diff
@@ -69,9 +69,23 @@
...
+    "bootstrap.servers": ",".join(DEFAULT_BROKERS),
+    "sasl.mechanism": os.environ.get("SASL_MECHANISM"),
+    "sasl.username": os.environ.get("SASL_PLAIN_USERNAME"),
+    "sasl.password": os.environ.get("SASL_PLAIN_PASSWORD"),
+    "security.protocol": os.environ.get("SECURITY_PROTOCOL"),
+}
+BROKER_CONFIG: Mapping[str, Any] = {
+    k: v for k, v in _BROKER_CONFIG.items() if v is not None
+}
+STORAGE_BROKER_CONFIG: Mapping[str, Mapping[str, Any]] = {}
```
[the remark this answers is not in the document]
Comment (alexef) [reply snuba/settings.py:87]: But as far as I understand the settings.py load order, at the time this execute, the use settings.py is not yet loaded. 

So my STORAGE_BROKER_CONFIG will always be empty, as DEFAULT_STORAGE_BROKERS is always empty in the default settings.py that is loaded before the local settings.py.

I can probably postpone the execution of the code that computes the value of STORAGE_BROKER_CONFIG  to the time it is first read, but it really doesn't feel like a proper solution.

What do you suggests with regards to load order?

[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/settings.py:87]: I think you are right, and this impacts DEFAULT_BROKER/BROKER_CONFIG. IF somebody overrides DEFAULT_BROKER, BROKER_CONFIG won't see the change. 
I guess there is no alternative other than keep the variables (DEFAULT_BROKER/BROKER_CONFIG and STORAGE_BROKER_CONFIG/DEFAULT_STORAGE_BROKERS) separate.

1) this PR would leave the old variables there empty
2) it would move the config in the new variables
3) the code (see the comment where I advise to move the config loading code in one place only) will read first the old variable, then, if empty the new one.
4) When enough time will have passed that we can decommission the old one we will remove that code_path.

File: snuba/settings.py [hunk 9]
```diff
@@ -69,9 +69,23 @@
...
 
+_BROKER_CONFIG: Mapping[str, Any] = {
+    "bootstrap.servers": ",".join(DEFAULT_BROKERS),
+    "sasl.mechanism": os.environ.get("SASL_MECHANISM"),
+    "sasl.username": os.environ.get("SASL_PLAIN_USERNAME"),
+    "sasl.password": os.environ.get("SASL_PLAIN_PASSWORD"),
+    "security.protocol": os.environ.get("SECURITY_PROTOCOL"),
+}
+BROKER_CONFIG: Mapping[str, Any] = {
+    k: v for k, v in _BROKER_CONFIG.items() if v is not None
```
[the remark this answers is not in the document]
Comment (alexef) [reply snuba/settings.py:85]: I will now do the function that loads the config, which strips out the None value, and remove this.

[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/settings.py:85]: Thanks. 
Could you please also validate in that function that the user is specifying parameters we support? This was the intention with the dataclass, but now we are not doing it anymore. A list of allowed parameters in the function is more than enough and you can start by filling it in with the ones you need. 
This would allow us to have a contract to what users may be overriding when we make changes to the config structure and still keep it very easy to support new ones. Making it more restrictive later for any reason would instead be very hard.
(raising an error if an unsupported parameter is found would be desirable).

**Comment by alexef** [comment]
@fpacifici ready for another round of review 🙏

File: snuba/utils/streams/backends/kafka.py [hunk 10]
```diff
@@ -615,26 +617,100 @@ def closed(self) -> bool:
...
+) -> KafkaBrokerConfig:
+    if storage_name in settings.DEFAULT_STORAGE_BROKERS:
+        # this is now deprecated
+        logger.warning(
+            "DEPRECATED: DEFAULT_STORAGE_BROKERS is defined. Please use STORAGE_BROKER_CONFIG instead"
+        )
+        default_config = {}
+        bootstrap_servers = ",".join(settings.DEFAULT_STORAGE_BROKERS[storage_name])
+    else:
+        default_config = settings.STORAGE_BROKER_CONFIG.get(
```
Comment (fpacifici) [inline snuba/utils/streams/backends/kafka.py:647]: I think this is missing the case where the user leaves `DEFAULT_STORAGE_BROKERS` empty and overrides `DEFAULT_BROKERS`. If the user does that we would end up using `BROKER_CONFIG` instead of `DEFAULT_BROKERS` which is wrong. 

I would advise you wrote a unit test for this logic. Overriding settings for unit tests is trivial and it would be very easy to ensure all cases are addressed in the right order:
- the user overrides one of the legacy variable
- the user overrides both legacy variables
- the user does not override legacy variables and relies on default
- the user does not override legacy variables but overrides the new ones

File: snuba/utils/streams/backends/kafka.py [hunk 11]
```diff
@@ -615,26 +617,100 @@ def closed(self) -> bool:
...
+)
+DEFAULT_STORAGE_NAME = ""
+
+
+def get_default_kafka_configuration(
+    storage_name: Optional[str] = None,
+    bootstrap_servers: Optional[Sequence[str]] = None,
+    override_params: Optional[Mapping[str, Any]] = None,
+) -> KafkaBrokerConfig:
+    if storage_name in settings.DEFAULT_STORAGE_BROKERS:
```
Comment (fpacifici) [inline snuba/utils/streams/backends/kafka.py:639]: Potentially someone could add a None key in the DEFAULT_STORAGE_BROKERS and that would be a valid key. So please don't look into DEFAULT_STORAGE_BROKERS and STORAGE_BROKER_CONFIG if the storage name you receive here is None. So we have a defined way of loading the default config where no storage is available.

Also please refer to storages with the `StorageKey` type instead of the bare string so mypy can type check this. 
```
storage_name: Optional[StorageKey] = None,
```

File: snuba/cli/replacer.py [hunk 12]
```diff
@@ -108,6 +105,7 @@ def replacer(
     replacer = StreamProcessor(
         KafkaConsumer(
             build_kafka_consumer_configuration(
+                storage_name,
```
Comment (fpacifici) [inline snuba/cli/replacer.py:108]: Provided details below. Please pass the storage key and not the storage name around.

File: snuba/consumers/consumer_builder.py [hunk 13]
```diff
@@ -39,7 +40,7 @@ def __init__(
         replacements_topic: Optional[str],
         max_batch_size: int,
         max_batch_time_ms: int,
-        bootstrap_servers: Sequence[str],
+        bootstrap_servers: Optional[str],
```
Comment (fpacifici) [inline snuba/consumers/consumer_builder.py:43]: You are still passing a sequence here in the consumer.py file so this does not type check now. I don't think you need this change at all.

**Review by fpacifici — CHANGES_REQUESTED** [review]
I think there is a bug in get_default_kafka_configuration. Please see my comments inline.

Other than that only minor changes.

File: snuba/utils/streams/backends/kafka.py (hunk 10 above)
Comment (alexef) [reply snuba/utils/streams/backends/kafka.py:647]: good call, yes let me add the unit test and fix this function

File: snuba/utils/streams/backends/kafka.py (hunk 11 above)
Comment (alexef) [reply snuba/utils/streams/backends/kafka.py:639]: will do

File: snuba/consumers/consumer_builder.py (hunk 13 above)
Comment (alexef) [reply snuba/consumers/consumer_builder.py:43]: it may be a left over from previous refactoring. let me double check

File: snuba/settings.py [hunk 14]
```diff
@@ -69,9 +69,21 @@
...
+# DEPRECATED, please use STORAGE_BROKER_CONFIG instead
 DEFAULT_STORAGE_BROKERS: Mapping[str, Sequence[str]] = {}
 
+BROKER_CONFIG: Mapping[str, Any] = {
+    # See snuba/utils/streams/backends/kafka.py for the supported options
+    "bootstrap.servers": os.environ.get("DEFAULT_BROKERS", "localhost:9092"),
+    "sasl.mechanism": os.environ.get("SASL_MECHANISM"),
+    "sasl.username": os.environ.get("SASL_PLAIN_USERNAME"),
+    "sasl.password": os.environ.get("SASL_PLAIN_PASSWORD"),
+    "security.protocol": os.environ.get("SECURITY_PROTOCOL"),
```
[the remark this answers is not in the document]
Comment (alexef) [reply snuba/settings.py:83]: I'll just drop them, no need to configure these over env

File: snuba/utils/streams/backends/kafka.py (hunk 10 above)
Comment (alexef) [reply snuba/utils/streams/backends/kafka.py]: updated and unittest added

File: snuba/consumers/consumer_builder.py [hunk 15]
```diff
@@ -84,13 +88,7 @@ def __init__(
...
         # XXX: This can result in a producer being built in cases where it's
         # not actually required.
-        self.producer = Producer(
-            {
-                "bootstrap.servers": ",".join(self.bootstrap_servers),
-                "partitioner": "consistent",
-                "message.max.bytes": 50000000,  # 50MB, default is 1MB
-            }
-        )
+        self.producer = Producer(self.broker_config)
```
Comment (fpacifici) [inline snuba/consumers/consumer_builder.py:91]: This will fail to apply the producer settings (partitioner and message.max.bytes) since those are applied by `build_kafka_producer_configuration` while`self.broker_config` is produced by `get_default_kafka_configuration`.
Please create two separate broker configs above in the constructor. One for consumers and the other for producers.

File: snuba/utils/streams/backends/kafka.py [hunk 16]
```diff
@@ -615,26 +618,120 @@ def closed(self) -> bool:
...
+                "DEPRECATED: DEFAULT_STORAGE_BROKERS is defined. Please use STORAGE_BROKER_CONFIG instead"
+            )
+            default_config = {}
+            default_bootstrap_servers = ",".join(
+                settings.DEFAULT_STORAGE_BROKERS[storage_name]
+            )
+        elif settings.DEFAULT_BROKERS:
+            logger.warning(
+                "DEPRECATED: DEFAULT_BROKERS is defined. Please use BROKER_CONFIG instead"
+            )
```
Comment (fpacifici) [inline snuba/utils/streams/backends/kafka.py:654]: Please remove this for now. This can be a lot of logs (it would be in at least one environment I know). I will add it back later after communicating these are deprecated in the next release and docs. You can leave the other one since almost none is using it.

File: snuba/utils/streams/backends/kafka.py [hunk 17]
```diff
@@ -615,26 +618,120 @@ def closed(self) -> bool:
...
+            default_bootstrap_servers = ",".join(settings.DEFAULT_BROKERS)
+        else:
+            default_config = settings.STORAGE_BROKER_CONFIG.get(
+                storage_name, settings.BROKER_CONFIG
+            )
+    else:
+        if settings.DEFAULT_BROKERS:
+            logger.warning(
+                "DEPRECATED: DEFAULT_BROKERS is defined. Please use BROKER_CONFIG instead"
+            )
```
Comment (fpacifici) [inline snuba/utils/streams/backends/kafka.py:665]: Same as above

File: snuba/utils/streams/backends/kafka.py [hunk 18]
```diff
@@ -615,26 +618,120 @@ def closed(self) -> bool:
...
+        else:
+            default_config = settings.BROKER_CONFIG
+    broker_config = default_config.copy()
+    bootstrap_servers = bootstrap_servers or default_bootstrap_servers
+    if bootstrap_servers:
+        broker_config["bootstrap.servers"] = bootstrap_servers
+    broker_config = {k: v for k, v in broker_config.items() if v is not None}
+    for configuration_key in broker_config:
+        if configuration_key not in SUPPORTED_KAFKA_CONFIGURATION:
+            raise RuntimeError(
```
Comment (fpacifici) [inline snuba/utils/streams/backends/kafka.py:677]: I think this is a good candidate for ValueError: https://docs.python.org/3/library/exceptions.html#ValueError

File: tests/utils/streams/backends/test_kafka_config.py [hunk 19]
```diff
@@ -0,0 +1,108 @@
+import importlib
+from snuba import settings
+from unittest import TestCase
+
+from snuba.datasets.storages import StorageKey
+from snuba.utils.streams.backends.kafka import get_default_kafka_configuration
+
+
+class TestKafkaConfig(TestCase):
```
Comment (fpacifici) [inline tests/utils/streams/backends/test_kafka_config.py:9]: nit: up to you, I would not block for this. It seems you are not gaining a lot by defining the test as a class. You can just add the test functions in this module without declaring the class. 
(Then you can implement the teardown this way: https://docs.pytest.org/en/2.8.7/xunit_setup.html#method-and-function-level-setup-teardown)

File: tests/utils/streams/backends/test_kafka_config.py [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+        settings.DEFAULT_STORAGE_BROKERS = {storage_name: [default_broker]}
+        broker_config = get_default_kafka_configuration(storage_key=storage_key)
+        self.assertEqual(broker_config["bootstrap.servers"], default_broker)
+
+        default_brokers = ["my.broker:9092", "my.second.broker:9092"]
+        settings.DEFAULT_STORAGE_BROKERS = {storage_name: default_brokers}
+        broker_config = get_default_kafka_configuration(storage_key=storage_key)
+        self.assertEqual(broker_config["bootstrap.servers"], ",".join(default_brokers))
+
+    def test_default_config_legacy_override_storage_servers_fallback(self):
```
Comment (fpacifici) [inline tests/utils/streams/backends/test_kafka_config.py:44]: excellent, thanks.

**Review by fpacifici — CHANGES_REQUESTED** [review]
Thanks.

This is really close. There is still one minor issue on the consumer_builder that needs to be fixed. Sorry, cannot without fixing that one.

File: snuba/utils/streams/backends/kafka.py [hunk 21]
```diff
@@ -615,26 +618,120 @@ def closed(self) -> bool:
...
+    else:
+        if settings.DEFAULT_BROKERS:
+            logger.warning(
+                "DEPRECATED: DEFAULT_BROKERS is defined. Please use BROKER_CONFIG instead"
+            )
+            default_config = {}
+            default_bootstrap_servers = ",".join(settings.DEFAULT_BROKERS)
+        else:
+            default_config = settings.BROKER_CONFIG
+    broker_config = default_config.copy()
```
Comment (fpacifici) [inline snuba/utils/streams/backends/kafka.py:670]: Please use a copy.deepcopy here. I believe there are a few settings that may be structured.

File: snuba/consumers/consumer_builder.py (hunk 15 above)
Comment (alexef) [reply snuba/consumers/consumer_builder.py]: you are right, I missed it. fixed now

**Comment by alexef** [comment]
I like how it looks now :)

**Review by fpacifici — COMMENTED** [review]
This looks really good. Thank you very much.
Will do a couple of last tests today and then merge.

File: snuba/utils/streams/backends/kafka.py [hunk 22]
```diff
@@ -615,26 +619,114 @@ def closed(self) -> bool:
...
+    else:
+        if settings.DEFAULT_BROKERS:
+            default_config = {}
+            default_bootstrap_servers = ",".join(settings.DEFAULT_BROKERS)
+        else:
+            default_config = settings.BROKER_CONFIG
+    broker_config = copy.deepcopy(default_config)
+    bootstrap_servers = bootstrap_servers or default_bootstrap_servers
+    if bootstrap_servers:
+        broker_config["bootstrap.servers"] = bootstrap_servers
```
Comment (fpacifici) [inline snuba/utils/streams/backends/kafka.py:668]: There is a bug here.
`default_bootstrap_servers` is a string here so everything is fine, though if you pass bootstrap-server to the command line. Now the line that joins the sequence into a string does not exist anymore.
It should be something like:
```
bootrstap_servers = ",".join(bootstrap_servers) if bootstrap_servers else default_bootstrap_servers. 
```

You can test this by removing the config from the settings and running a consumer this way:
```
snuba consumer --storage=events --bootstrap-server=localhost:9092
```

Now the confluent Kafka API receives a sequence instead of a string.

**Review by fpacifici — CHANGES_REQUESTED** [review]
Sorry I found another issue. The command line override does not work properly, this should be an easy case to be added to the unit test.

File: snuba/utils/streams/backends/kafka.py (hunk 22 above)
Comment (alexef) [reply snuba/utils/streams/backends/kafka.py:668]: Good catch. Let me add a test case and fix this!

Comment (alexef) [reply snuba/utils/streams/backends/kafka.py]: updated

File: snuba/state/__init__.py [hunk 23]
```diff
@@ -214,7 +217,7 @@ def record_query(query_metadata: Mapping[str, Any]) -> None:
         ).execute()
 
         if kfk is None:
-            kfk = Producer({"bootstrap.servers": ",".join(settings.DEFAULT_BROKERS)})
+            kfk = Producer(build_default_kafka_producer_configuration())
```
Comment (fpacifici) [inline snuba/state/__init__.py:220]: There is a bug here, this does not behave as before. By using `build_default_kafka_producer_configuration` you are calling into `build_kafka_producer_configuration` which applies these two parameters which are not default:
```
"partitioner": "consistent",
"message.max.bytes": 50000000,  # 50MB, default is 1MB
```

You may want to remove those two default parameters from `build_kafka_producer_configuration` and always rely on overrides to set them when needed since they are different from the librdkafka defaults:
https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md

**Review by fpacifici — CHANGES_REQUESTED** [review]
Sorry one more to fix. This way the producer in state/__init__.py would not behave as in master.

File: snuba/state/__init__.py (hunk 23 above)
Comment (alexef) [reply snuba/state/__init__.py:220]: Fair enough. Updated to remove default parameters.

**Review by fpacifici — APPROVED** [review]
I think this is ready.
Thanks a lot.

**Comment by fpacifici** [comment]
Ok, there is an issue with the cloudbuild config, I should be able to fix it tomorrow morning and then merge.

**Comment by olliezhu** [comment]
@alexef @fpacifici Do any changes need to be made to fully support TLS auth with kafka? Looks like we are only adding SASL fields to the configuration here although the PR title mentions TLS. What kinds of changes need to be added to enable (for example) mTLS besides enabling configuration for fields like `ssl.truststore`, `ssl,keystore`, `ssl.key`, etc.?

**Comment by fpacifici** [comment]
Thanks to @alexef work you should be able to add new allowed configuration parameters by simply adding them to this list:
For snuba:
https://github.com/getsentry/snuba/blob/master/snuba/utils/streams/backends/kafka.py#L624-L632
For Sentry:
https://github.com/getsentry/sentry/blob/master/src/sentry/utils/kafka_config.py#L5-L16

The ssl parameters are no brainers, there will most likely be no discussion on whether they should be supported or not. So please feel free to send a PR with the changes and the review will be quick.

Then you should be able to make the settings changes.

**Comment by bumarcell** [comment]
Hello Everybody
this thread is supposedly merged but I cannot find a way to find/use those changes. 

Did I miss something?

**Comment by ChannyClaus** [comment]
> Hello Everybody this thread is supposedly merged but I cannot find a way to find/use those changes.
> 
> Did I miss something?

_i believe_ https://github.com/getsentry/snuba/blob/8304cc553d55f4f40ed3049887430440e3e4c163/snuba/settings/__init__.py#L233 is the way to use but i haven't touched this in a while so someone correct me if i'm wrong

**Comment by fpacifici** [comment]
Hi, 
sorry for the late answer. 

@ChannyClaus is right. 
The way to set specific  settings for kafka are these settings:
https://github.com/getsentry/snuba/blob/master/snuba/settings/__init__.py#L115-L128

- `BROKER_CONFIG` defines the parameters for the default broker.
- `KAFKA_BROKER_CONFIG` allows you to override the parameters for specific topics by assigning topics to different brokers. If you have all your topics on the same broker, ignore this.

The list of allowed settings though is not in the Snuba code base (probably we should fix that).
It is in the streaming library: https://github.com/getsentry/arroyo/blob/main/arroyo/backends/kafka/configuration.py#L16-L38
We can add new fields if needed. We keep them based on an allow list to make breaking changes less likely.

**Comment by bumarcell** [comment]
So I pulled the repo and added `ssl.ca.pem`, `ssl.certificate.pem` and `ssl.key.pem` as following:
image
Then built the image and deployed it. Then I got the following error:
image

Did I miss something there? 🤔

**Comment by bumarcell** [comment]
Ok I had to use `ssl.ca.location` and it worked 👍 
One problem tho is that all topics are being written on only one replica 😞
